### PR TITLE
perf(storage): reduce calling async method with poll_next in iterator

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

In previous benchmark, we have observed that calling too many async method may cause performance regression even if all types are static. Besides, we observe that in most call on `next`, we don't need to call async method and await. In most cases it's just advancing a logical pointer without any necessity to await on anything. Therefore, in this PR, we split the frequently called method `next` into two parts, `poll_next` and `await_next`. The `poll_next` will do all the non-async stuffs, and therefore it's not an async method, and the remaining `async` stuffs in `next` will be done in `await_next`. The `poll_next` method will return a `core::task::poll` enum, where `Poll::Ready` means that the `next` can be finished without calling `await_next`, and `Poll::Pending` means that we should further call `await_next` to finish the remaining async stuff.

The `next` will become
```
fn next(&mut self) -> Self::NextFuture<'_> {
        async {
            match self.poll_next() {
                Poll::Ready(result) => result,
                Poll::Pending => self.await_next().await,
            }
        }
    }
```

We run the benchmark `bench_compactor` and observe great improvement.
Main:
image

Current branch:
image

The merge iterator using the normal concat sstable iterator wrapped by union has 35% improvement. 
The merge iterator using the normal concat sstable iterator has 21% improvement.
The merge iterator using the the sstable-prefetched iterator has 5% improvement.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/storage/src/hummock/iterator/concat_inner.rs [hunk 1]
```diff
@@ -100,21 +112,65 @@ impl<TI: SstableIteratorType> ConcatIteratorInner<TI> {
...
+                                Ok(())
+                            } else {
+                                self.seek_idx(self.cur_idx + 1, None).await
+                            }
+                        }
+                        Err(e) => Err(e),
+                    }
+                }
+                ConcatIteratorPendingStage::AwaitSeekIdx => {
+                    self.seek_idx(self.cur_idx + 1, None).await
```
Comment (skyzh) [inline src/storage/src/hummock/iterator/concat_inner.rs:169]: You should either implement the whole function as async state machine or using async function. I rarely see people mix these two. Not sure whether there will be incorrectness from such implementation.

File: src/storage/src/hummock/iterator/merge_inner.rs [hunk 2]
```diff
@@ -183,93 +188,100 @@ where
...
-            // before returning.
+        // WARNING: within scope of BinaryHeap::PeekMut, we must carefully handle all places of
+        // return. Once the iterator enters an invalid state, we should remove it from heap
+        // before returning.
 
-            // Put the popped nodes back to the heap if valid or unused_iters if invalid.
-            for mut node in popped_nodes {
-                match node.iter.next().await {
+        // Put the popped nodes back to the heap if valid or unused_iters if invalid.
+        for mut node in popped_nodes {
```
Comment (skyzh) [inline src/storage/src/hummock/iterator/merge_inner.rs:217]: This state machine doesn't look equivalent to the previous one. For this one, if there's any iterator pending, we will go over the `while let Some(next_node) = self.heap.peek_mut()` loop when calling `poll_next_inner`.

**Comment by BugenZhao** [comment]
The `poll_next` naming seems confusing to the real `Future::poll_next`. :(

By the way, if there's no `dyn Future` on the path, I think the compiler is able to realize that and the generated state machine should be the same as calling a sync function.

File: src/storage/src/hummock/iterator/merge_inner.rs (hunk 2 above)
Comment (wenym1) [reply src/storage/src/hummock/iterator/merge_inner.rs:217]: Before we call this method we should ensure that there is no pending iterator. We should call `await_next` after Poll::Pending is returned from `poll_next`, and the `await_next` should have awaited all pending iterators.

The usage of `poll_next` is different to the conventional usage of `poll` in that when `Pending` is returned, we should not call `poll_next` again, and should call `await_next`.

**Comment by skyzh** [comment]
> The `poll_next` naming seems confusing to the real `Future::poll_next`. :(
> 
> By the way, if there's no `dyn Future` on the path, I think the compiler is able to realize that and the generated state machine should be the same as calling a sync function.

Yes, I just realized that. This PR is not implementing Future. It just splits the non-async thing out of the original function. I would prefer not naming it as `poll_next`, and don't use `std::task::Poll` here to avoid confusion.

**Comment by wenym1** [comment]
> By the way, if there's no dyn Future on the path, I think the compiler is able to realize that and the generated state machine should be the same as calling a sync function.

I used to think so. But some benchmark result is really confusing, and it seems that we can get great improvement if we avoid using await. I suspect that async method can cause some regression for compiler code optimization.

**Comment by wenym1** [comment]
> I would prefer not naming it as poll_next, and don't use std::task::Poll here to avoid confusion.

Do you have any suggestion on the naming? I have been thinking about it the first day I wrote this PR 🤣

**Comment by skyzh** [comment]
Initially I guess the compiler somehow cannot do LTO without this PR, but I'm wrong. I removed these two lines from `Cargo.toml`.

```
codegen-units = 1
lto = 'thin'
```

After this PR, with LTO
```
bench_union_merge_iterator                                                                            
                        time:   [90.442 ms 90.546 ms 90.663 ms]
```

Before this PR, with LTO

```
bench_union_merge_iterator                                                                            
                        time:   [111.53 ms 111.67 ms 111.84 ms]
                        change: [+23.123% +23.331% +23.564%] (p = 0.00 < 0.05)
```

Before this PR, w/o LTO

```
bench_union_merge_iterator                                                                            
                        time:   [149.62 ms 149.88 ms 150.15 ms]
```

After this PR, w/o LTO

```
bench_union_merge_iterator                                                                            
                        time:   [127.23 ms 127.52 ms 127.81 ms]
```

So even w/o LTO, we can get performance improvement with this PR.


One thing to notice is that we currently don't enable LTO in release builds. With LTO we can get 149.88 ms -> 111.67 ms = 25% improvement. So LTO can already do a lot, we can enable LTO in release builds to have better performance. For this part, I would prefer making the code clear to read to the 15% further improvement.

**Comment by BugenZhao** [comment]
> > The `poll_next` naming seems confusing to the real `Future::poll_next`. :(
> > By the way, if there's no `dyn Future` on the path, I think the compiler is able to realize that and the generated state machine should be the same as calling a sync function.
> 
> Yes, I just realized that. This PR is not implementing Future. It just splits the non-async thing out of the original function. I would prefer not naming it as `poll_next`, and don't use `std::task::Poll` here to avoid confusion.

+1. 🥵

**Comment by Little-Wallace** [comment]
How about use `poll_next` in `compact` ?
For example:

```rust
async fn compact(...) {
    while iter.is_valid() {
        iter.poll_next();
        if iter.is_pending() {
            iter.await_next().await;
        }
    }
}

```

**Comment by wenym1** [comment]
> How about use poll_next in compact ?
For example:
```
async fn compact(...) {
    while iter.is_valid() {
        iter.poll_next();
        if iter.is_pending() {
            iter.await_next().await;
        }
    }
}
```
******

The `next` of iter in this PR is already similar to
```
async fn next() {
    iter.poll_next();
    if iter.is_pending() {
        iter.await_next().await;
    }
}
```

So the mentioned
```
async fn compact(...) {
    while iter.is_valid() {
        iter.poll_next();
        if iter.is_pending() {
            iter.await_next().await;
        }
    }
}
```
will be almost the same as 
```
async fn next() {
    while iter.is_valie() {
        iter.next().await
    }
}
```
except for one more layer of await called in `compact`.

**Comment by Little-Wallace** [comment]
But the previous method would not generate a structure `NextFuture`....
I think this structure NextFuture will be much heavy than `Poll`

**Pull request closed without merging**